### ISSUE TITLE
Improper frame pacing (synchronization) on Flycast standalone

### ISSUE THREAD

**Issue by HelicopterP** [flyinghead/flycast#1733]
Platform / OS / Hardware: Win11 i7 12700 RTX3070 160Hz VRR display 

Flycast version: 2.4 Standalone

**_Description of the Issue_**

Standalone Flycast versions do not sync framerate properly which results in unpleasant stuttery presenatation of every piece of software running on it. Retroarch versions are fine becuase RA have an option to sync refresh rate to exact content framerate on VRR screens. Standalone Flycast on the other hand do not allow to do this no matter the settings - Vsync on/off, Fullscreen/windowed, Doubling framerate on/off, etc. I am pretty sure that the cause of this issue is synchronization because RA with Vsync off and VRR disabled looks pretty similar to Flycast standalone (stuttery). 

**_Debugging Steps Tested_**

Launch 240p test suite, select video tests then scroll test. Observe stuttery and uneven scrolling. Uneven scrolling is present in every game but it is easily visible in 240p test suite. RA with sync refresh rate to exact content framerate on is fine on the other hand.

**Comment by HelicopterP** [comment]
It won't help. Maybe you are not sentive to inconsistent frametimes so here is visual representation:
This is standalone with Vsync on 60Hz display:
![standalone 60Hz](https://github.com/user-attachments/assets/e46177be-a43a-4423-8264-4ea20027f1f7)
This is standalone with Vsync on 59.94Hz diplay:
![standalone 59 94Hz](https://github.com/user-attachments/assets/782b6aad-bcd5-42d9-a008-c1be0f747a0f)
Please pay atention to frametime and FPS graphs - both of them are all over the place.

And here is Retroarch with Vsync and VRR on on 160Hz display:
![Retroarch 160Hz](https://github.com/user-attachments/assets/b50bdeeb-a89f-4855-9088-c46435ff249f)
While frametime is a bit wobbly it is so so much better than a standalone version. Yes technically Retroarch version is not perfect too but it is miles above standalone version. To be honest in this kind of scroll test standalone is completely broken

**Comment by flyinghead** [comment]
I notice 1 stutter/frame drop every 5 or 10 secs on the horizontal scroll test.
Is it what you're seeing?

**Comment by HelicopterP** [comment]
>I notice 1 stutter/frame drop every 5 or 10 secs on the horizontal scroll test.
This is my experience with Retroarch version -not ideal but ok.
Standalone is a mess on the other hand. There are multiple duplicate frames during any given second. Here is a video https://drive.google.com/file/d/1PYs-hYM15ElvU5730omZL1v5GLdWBdfu/view?usp=sharing
You can advance frame by frame with E key in VLC for example. 
BTW, this is not unique to my PC only. I experinced the same problem on Win10 and GTX1070 for example.

**Comment by flyinghead** [comment]
Yes, there is clearly something wrong with your setup or configuration. It looks like VSync is off or the monitor isn't in fixed 60 Hz.

**Comment by HelicopterP** [comment]
That recording was done with monitor set to 160Hz mode but refreshrate of the monitor doesn't change a thing in this case. Here is a video made with a monitor set to 60Hz - it looks a bit different but the core problem is the same. https://drive.google.com/file/d/1_athr7J08gYSbkjMjFfOl6ZoupVOZPOn/view?usp=sharing
Vsync is most definitely enabled because without it I get constant screen tearing in the middle of the screen and stuttering doesn't go anywhere. In any case requiring a 60Hz-refresh rate of the monitor kinda implies that something is amiss anyway especially in 2024.

I can asure you that the issue is real and not unique to my PC only. As I said earlier I noticed it like 2 years ago on different hardware on an older version of Flycast and the problem was the same. I also tried it with default settings (deleting .cfg file), different APIs, different refresh rates, Gsync on/off in driver settings, etc and issue is still there. I am also sure that many people simply do not notice this because standalone version is less popular, no one plays 240p test suite and while stuttering is visible in any content it is harder to detect in non scrolling games. If this would've been a Genesis emulator you would've heard a lot more complaints.

flyinghead, what is your hardware. Maybe you are on Linux and AMD?

**Comment by flyinghead** [comment]
I develop mostly on ubuntu 20 with a Nvidia GTX 1080.
But here is a video recorded on a low end windows 10 system with integrated Iris+ 640 GPU.

https://github.com/user-attachments/assets/a0798e1c-b4f6-4880-b159-dc65670b4e0b

**Comment by arcanechocolate** [comment]
@HelicopterP This issue has been fixed in https://github.com/blueminder/flycast-dojo/.

**Comment by HelicopterP** [comment]
> @HelicopterP This issue has been fixed in https://github.com/blueminder/flycast-dojo/.

Indeed, setting Fixed Frequency (for 240 test suite) to auto or 60Hz or 59.94Hz (this one works because of VRR I guess) fixes this issue. Setting it to 60Hz is similar to PCSX2 and Duckstation option "Sync to Host Refresh rate" I presume, which is good for non-VRR screens.

But this lead me to dig deeper and here is what I found.  Flycast dojo have an option called "Audio Frame Sync". Ticking it makes it behave like an original Flycast. So I thought to try it with sound disabled in Windows and ... it fixes stuttering!


> I develop mostly on ubuntu 20 with a Nvidia GTX 1080. But here is a video recorded on a low end windows 10 system with integrated Iris+ 640 GPU.

The only way for me to replicate this result is to run flycast on system with no audio devices or audio devices disabled. I suppose this is where problem lies - something is going wrong while trying to sync with audio. I've tried to play with different out sound quality options in Windows but the issue remains.

**Comment by HelicopterP** [comment]
After further digging I am pretty sure that, yep, current synchronization method is not good on many-many levels.
1. Setting audio driver to null causes 30FPS games to speed up to 60fps (everything is running 2 times faster than normal) but... it improves frame pacing on my system considerably. Any other audio driver causes stuttering.
2. Unpluging _the only_ audio device (headphones for example) during emulation causes flycast to hang.
3. Launching a game in flycast with no audio devices connected to PC takes a long time.
4. Flycast standalone does not trigger VRR mode on displays that support it. Windowed or fullscreen - flycast outputs whatever it wants but monitor refreshrate stays at 160Hz (or whatever your monitor is set to). Setting display to 60Hz or 59.94Hz doesn't fix anything as flycast output does not match this framerates exactly - this causes one skiped frame per several seconds (assuming it doesn't stutter all the time - I can only achive it with audio disabled one way or the other). Flycast-dojo on the other hand supports VRR, so both 60Hz and 59.94Hz modes are stutter-free on VRR display (with audio synchronization disabled).

The issue is very old and not unique to me. This a random video of Sega GT running on flycast that I found on youtube https://www.youtube.com/watch?v=cEcHTeeFpwQ&t=173s - at 2.54 there are several panning camera shots and all of them are a stuttering mess. I was unable to find a recording of the exact track from a real DC, but here is another track https://www.youtube.com/watch?v=ikPQeahsfn0&t=105s and it is alot smoother. I would presume that small stutter we see here is because of 60 Hz-59.94Hz mismatch, anyway both this recording and real DC are a lot smoother than flycast.

_tl;dr Audio synchronization does not work properly and even when it does flycast does not trigger VRR mode on supported displays which result in stutter._

flyinghead, I am curious why was flycast designed this way? I mean why is video being synchronized to audio and not the other way around like other emulators do? I would guess, that having two modes (60fps - for non-VRR screens, 59.94fps for VRR screens (and triggering VRR) while correctly handling 30fps inside of them) and synchronizing audio to video would have solved all of the above problems.

**Comment by flyinghead** [comment]
Flycast normally uses 2 synchronization sources:
1. Audio output (the emulator will wait when the audio output buffer is full)
2. VSync (the emulator will wait for VSync to present its frame)

Without these, flycast will run at maximum speed, like fast forwarding.

1. By setting the audio driver to null, you disable audio output synchronization. So the emulator will only use VSync, and if running a 30 FPS game, it will run it at x2 speed.
2. Not sure how this is related to frame pacing
4. Again, unrelated. If you want to report these audio problems, please do so in a different issue.
5. VRR isn't supported. One reason being that Flycast will NEVER output more than 60 frames/sec.

Sega GT is a 30 FPS game and there is a known issue with 30 FPS games. See #520. However this is unrelated to VRR and VSync issues.

If you are curious, Flycast is open source and you can look at the code and even propose changes. You seem to know how things *should* be done.
